### DISCUSSION TITLE
Co-existence of Vulkan and VulkanSC headers

### DISCUSSION THREAD

**Discussion by mgorchak-blackberry** [KhronosGroup/VulkanSC-Docs#7] [General]
Hi,

I have a question about co-existence of Vulkan and VulkanSC headers installed to the same include/vulkan location. I've found that vk_icd.h, vk_layer.h, most WSI headers are incompatible, because of generation using different path in vk.xml . How this could be resolved by OS vendor who needs to ship both Vulkan and VulkanSC headers?

Thanks!

**Comment by dgkoch** [comment]
Hi @mgorchak-blackberry,

On a fully "safety certified" system, we don't think this would be an issue, as you wouldn't actually include Vulkan on that platform. 

However, having both is obviously very useful for a development environment, or even an environment where Vulkan SC is desired to be used for production but of non-safety certified use cases (which we have envisioned for a variety of use cases). 

NVIDIA does have systems with both Vulkan SC and Vulkan on them (see https://developer.nvidia.com/blog/using-vulkan-sc-for-safety-critical-graphics-and-real-time-gpu-processing/) 
We include an extra "VulkanSC" in the path for the Vulkan SC headers. 

For example if we have headers installed by a development SDK we would have:

- <sdk_path>/include/VulkanSC/vulkan/ (vulkan sc headers)
- <sdk_path>/include/vulkan/ (vulkan headers)

And then an application can change their include path at build time by using `-I <sdk_path>/include` or `-I <sdk_path>/include/VulkanSC` as long as the application includes the appropriate header from the sources using `#include <vulkan/vk_layer.h>` or whatever. Obviously the app will need to explicitly change includes for `vulkan_core.h` to `vulkan_sc_core.h` as well.

**Reply by mgorchak-blackberry** [reply]
Thank you very much! With VulkanSC path prefix it will work for QNX and fully resolves the problem.

One more small thing to confirm, VulkanSC ICD loader should have this name libvulkansc.so.XXXX ? So ICD loaders could co-exist too on the development host. 

I'm not sure if this is the right topic. If I understand things right - the one of the design ideas of VulkanSC is to not have WSI what demands Windowing System buffers allocation in runtime, instead pre-allocated buffers must be wrapped as Vulkan images and used as render targets. So the question is application must be responsible for buffers management and displaying them using Windowing System? As far as I understand nVidia SCI was designed exactly to do this.

Thank you!

**Comment by dgkoch** [comment]
> One more small thing to confirm, VulkanSC ICD loader should have this name libvulkansc.so.XXXX ? So ICD loaders could co-exist too on the development host.

This is correct.  When you build https://github.com/KhronosGroup/VulkanSC-Loader in Vulkan SC mode (-D VULKANSC=ON) it should create libvulkansc.so. 

> the one of the design ideas of VulkanSC is to not have WSI what demands Windowing System buffers allocation in runtime, instead pre-allocated buffers must be wrapped as Vulkan images and used as render targets. So the question is application must be responsible for buffers management and displaying them using Windowing System? As far as I understand nVidia SCI was designed exactly to do this.

WSI might warrant it's own discussion thread, but I'll try to do a quick reply here. 
I would say your above statement is partially correct - we don't want WSI for Vulkan SC to require allocating buffers during normal runtime. But this is largely true in Vulkan today - the only time you need to recreate swapchains is when resizing windows or when other window-system dependent properties change.  We don't think that anyone writing a true safety critical application is going to want to use a regular windowing system like win32, X, Wayland etc, just because it would be impractical to consider safety certifying the window system itself, and that's why we haven't chosen to expose the "regular" WSI extensions in Vulkan SC. Of course that could be useful in a development environment, but so far we haven't seen a compelling enough need for that. We are of course open to being proven wrong :-)

There are two ways that we see WSI being done in Vulkan SC. 
1. Use the "direct-to-display" extensions, VK_KHR_display, VK_KHR_display_swapchain, etc.  When these extensions are used, the Vulkan (SC) application typically has complete control of the display, and there is no window system running. If there are multiple applications that want to share the display, one could envision a Vulkan SC-base compositor application that received images from multiple applications (possibly via the mechanism in my next point) and arranged to display them as desired on that display. 
2. Use the standard Vulkan interop mechanisms (memory and sync primitives) to share the display image from the Vulkan SC application with another component on the system which is responsible for displaying the image (possibly with composition). 

NVIDIA's NVSci extensions are a form of (2) interop where the NvSciBuf objects are imported into Vulkan SC and then used as memory for backing the display image in VKSC, and then the shared image is presented via OpenWFD.

**Comment by mgorchak-blackberry** [comment]
Thanks a lot for your answers!
